### PR TITLE
subscription model feature

### PR DESCRIPTION
**Pull request by jenananthan**

### CODE REVIEW

File: features/org.wso2.carbon.appmgt.core.feature/src/main/resources/sql/mysql.sql [hunk 1]
```diff
@@ -240,6 +240,24 @@ CREATE TABLE IF NOT EXISTS APM_EXTERNAL_STORES (
     PRIMARY KEY (APP_STORE_ID)
 )ENGINE INNODB;
 
+CREATE TABLE IF NOT EXISTS APM_FAVOURITE_APPS (
+   ID INTEGER  AUTO_INCREMENT,
+   USER_ID VARCHAR(50) NOT NULL,
+   TENANT_ID INTEGER NOT NULL,
+   APP_ID INTEGER NOT NULL,
+   ADDED_TIME TIMESTAMP NOT NULL,
```
Comment (sajithab) [inline features/org.wso2.carbon.appmgt.core.feature/src/main/resources/sql/mysql.sql:248]: Better to set default time stamp as current time stamp.

File: features/org.wso2.carbon.appmgt.core.feature/src/main/resources/sql/mysql.sql [hunk 2]
```diff
@@ -240,6 +240,24 @@ CREATE TABLE IF NOT EXISTS APM_EXTERNAL_STORES (
     PRIMARY KEY (APP_STORE_ID)
 )ENGINE INNODB;
 
+CREATE TABLE IF NOT EXISTS APM_FAVOURITE_APPS (
+   ID INTEGER  AUTO_INCREMENT,
+   USER_ID VARCHAR(50) NOT NULL,
+   TENANT_ID INTEGER NOT NULL,
```
Comment (sajithab) [inline features/org.wso2.carbon.appmgt.core.feature/src/main/resources/sql/mysql.sql:246]: USER_ID + TENANT_ID -> UNIQUE

File: features/org.wso2.carbon.appmgt.core.feature/src/main/resources/sql/mysql.sql (hunk 1 above)
Comment (sajithab) [inline features/org.wso2.carbon.appmgt.core.feature/src/main/resources/sql/mysql.sql:248]: ADDED_TIME better to rename as CREATED_TIME

File: components/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/APIConsumer.java [hunk 3]
```diff
@@ -324,4 +325,94 @@ public String addApplication(Application application, String userId) throws
...
+     * @param tenantIdOfUser
+     * @param tenantIdOfStore
+     * @return true if favourite app else false
+     * @throws AppManagementException
+     */
+    public boolean isFavouriteApp(APIIdentifier identifier, String userId, int tenantIdOfUser, int tenantIdOfStore)
+            throws AppManagementException;
+
+    /**
+     * This methods returns  favourite webapps of given user for given tenant store.
```
Comment (jenananthan) [inline components/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/APIConsumer.java:367]: just tell return favourite webapps _>

File: components/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/APIConsumer.java [hunk 4]
```diff
@@ -324,4 +325,94 @@ public String addApplication(Application application, String userId) throws
...
+     * @param tenantIdOfStore @throws AppManagementException
+     */
+    public void removeFavouritePage(String userName, int tenantIdOfUser, int tenantIdOfStore)
+            throws AppManagementException;
+
+    /**
+     * This method checks whether user has  selected favourite page as default landing page in store.
+     *
+     * @param userName        User Name
+     * @param tenantIdOfUser
```
Comment (tgtshanika) [inline components/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/APIConsumer.java:413]: add parameter description of tenantIdOfUser and tenantIdOfStore

File: components/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/APIConsumer.java [hunk 5]
```diff
@@ -324,4 +325,94 @@ public String addApplication(Application application, String userId) throws
...
+    public void setFavouritePage(String userName, int tenantIdOfUser, int tenantIdOfStore)
+            throws AppManagementException;
+
+    /**
+     * This method mark that user has deselected the favourite page from default landing page in store.
+     *
+     * @param userName        User Name
+     * @param tenantIdOfUser
+     * @param tenantIdOfStore @throws AppManagementException
+     */
```
Comment (jenananthan) [inline components/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/APIConsumer.java:405]: complete method comments

File: components/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/APIConsumer.java [hunk 6]
```diff
@@ -324,4 +325,94 @@ public String addApplication(Application application, String userId) throws
...
+     * @throws AppManagementException
+     */
+    public List<APIIdentifier> getUserSubscribedApps(String userName, int tenantIdOfUser, int tenantIdOfStore)
+            throws AppManagementException;
+
+    /**
+     * This method mark that user has selected favourite page as default landing page in store.
+     *
+     * @param userName        User Name
+     * @param tenantIdOfUser  Tenant Id Of User
```
Comment (tgtshanika) [inline components/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/APIConsumer.java:392]: User Name --> username

File: components/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/APIConsumer.java [hunk 7]
```diff
@@ -324,4 +325,94 @@ public String addApplication(Application application, String userId) throws
...
+     * @throws AppManagementException
+     */
+    public void addToFavouriteApps(APIIdentifier identifier, String userId, int tenantIdOfUser, int tenantIdOfStore)
+            throws AppManagementException;
+
+    /**
+     * Remove the given app from favourite app list of given user.
+     *
+     * @param identifier      App identifier
+     * @param userId          User Id(user name)
```
Comment (tgtshanika) [inline components/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/APIConsumer.java:344]: user id username brings same meaning. So refactor user id --> username

File: components/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/model/WebApp.java [hunk 8]
```diff
@@ -86,6 +86,7 @@
     private String oldOutSequence;
 
     private boolean advertiseOnly;
+    private String advertisedAppUuid;
```
Comment (sajithab) [inline components/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/model/WebApp.java:89]: advertisedAppUuid -> advertisedAppUUID

File: components/org.wso2.carbon.appmgt.hostobjects/src/main/java/org/wso2/carbon/appmgt/hostobjects/APIStoreHostObject.java [hunk 9]
```diff
@@ -3164,7 +3165,443 @@ public static boolean jsFunction_isTenantActive(Context cx, Scriptable thisObj,
                     "is active or not", e);
         }
 
+    }
+
+
+    /**
+     * Add the given app to favourite app list of user for given tenant store.
```
Comment (jenananthan) [inline components/org.wso2.carbon.appmgt.hostobjects/src/main/java/org/wso2/carbon/appmgt/hostobjects/APIStoreHostObject.java:3172]: Add the given app to favourite app list of the user for a given tenant store

File: components/org.wso2.carbon.appmgt.hostobjects/src/main/java/org/wso2/carbon/appmgt/hostobjects/APIStoreHostObject.java [hunk 10]
```diff
@@ -3164,7 +3165,443 @@ public static boolean jsFunction_isTenantActive(Context cx, Scriptable thisObj,
...
+            apiConsumer.addToFavouriteApps(apiIdentifier, userName, tenantIdOfUser, tenantIdOfStore);
+        } finally {
+            if (isTenantFlowStarted) {
+                PrivilegedCarbonContext.endTenantFlow();
+            }
+        }
+    }
+
+    /**
+     * Remove the given app from favourite app list of user for given tenant store.
```
Comment (jenananthan) [inline components/org.wso2.carbon.appmgt.hostobjects/src/main/java/org/wso2/carbon/appmgt/hostobjects/APIStoreHostObject.java:3220]: check grammer

File: components/org.wso2.carbon.appmgt.hostobjects/src/main/java/org/wso2/carbon/appmgt/hostobjects/APIStoreHostObject.java [hunk 11]
```diff
@@ -3164,7 +3165,443 @@ public static boolean jsFunction_isTenantActive(Context cx, Scriptable thisObj,
...
+        try {
+            String tenantDomain = tenantDomainOfStore;
+            if (tenantDomain != null && !MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain)) {
+                isTenantFlowStarted = true;
+                PrivilegedCarbonContext.startTenantFlow();
+                PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantDomain(tenantDomain, true);
+            }
+
+            int tenantIdOfStore = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantId();
+            APIIdentifier apiIdentifier = new APIIdentifier(providerName, apiName, version);
```
Comment (sajithab) [inline components/org.wso2.carbon.appmgt.hostobjects/src/main/java/org/wso2/carbon/appmgt/hostobjects/APIStoreHostObject.java:3209]: api\* -> use as app\* in all the places.

File: components/org.wso2.carbon.appmgt.hostobjects/src/main/java/org/wso2/carbon/appmgt/hostobjects/APIStoreHostObject.java [hunk 12]
```diff
@@ -3164,7 +3165,443 @@ public static boolean jsFunction_isTenantActive(Context cx, Scriptable thisObj,
...
+     * @param thisObj Scriptable object
+     * @param args    Passing arguments
+     * @param funObj  Function object
+     * @return
+     * @throws ScriptException
+     * @throws AppManagementException
+     */
+    public static NativeArray jsFunction_getAllFavouriteApps(Context cx,
+                                                             Scriptable thisObj, Object[] args, Function funObj)
+            throws ScriptException, AppManagementException {
```
Comment (sajithab) [inline components/org.wso2.carbon.appmgt.hostobjects/src/main/java/org/wso2/carbon/appmgt/hostobjects/APIStoreHostObject.java:3329]: ScriptException ??

File: components/org.wso2.carbon.appmgt.hostobjects/src/main/java/org/wso2/carbon/appmgt/hostobjects/APIStoreHostObject.java [hunk 13]
```diff
@@ -3164,7 +3165,443 @@ public static boolean jsFunction_isTenantActive(Context cx, Scriptable thisObj,
...
+     * This method return the details of subscribed apps of given user for given tenant store.
+     *
+     * @param cx      Rhino context
+     * @param thisObj Scriptable object
+     * @param args    Passing arguments
+     * @param funObj  Function object
+     * @return
+     * @throws AppManagementException
+     */
+    public static NativeArray jsFunction_getUserSubscribedApps(Context cx,
```
Comment (jenananthan) [inline components/org.wso2.carbon.appmgt.hostobjects/src/main/java/org/wso2/carbon/appmgt/hostobjects/APIStoreHostObject.java:3405]: change name to getSubscriptionsOnTenantStore

File: components/org.wso2.carbon.appmgt.hostobjects/src/main/java/org/wso2/carbon/appmgt/hostobjects/APIStoreHostObject.java [hunk 14]
```diff
@@ -3164,7 +3165,443 @@ public static boolean jsFunction_isTenantActive(Context cx, Scriptable thisObj,
...
+        try {
+            String tenantDomain = tenantDomainOfStore;
+            if (tenantDomain != null && !MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain)) {
+                isTenantFlowStarted = true;
+                PrivilegedCarbonContext.startTenantFlow();
+                PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantDomain(tenantDomain, true);
+            }
+            int tenantIdOFStore = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantId();
+            APIConsumer apiConsumer = getAPIConsumer(thisObj);
+            apiConsumer.setFavouritePage(userName, tenantIdOfUser,tenantIdOFStore);
```
Comment (tgtshanika) [inline components/org.wso2.carbon.appmgt.hostobjects/src/main/java/org/wso2/carbon/appmgt/hostobjects/APIStoreHostObject.java:3513]: format code

File: components/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/dao/AppMDAO.java [hunk 15]
```diff
@@ -5687,22 +5687,50 @@ public APIKeyValidationInfoDTO getApplicationData(String appContext, String appV
...
+            boolean isEnterpriseSubscriptionEnabled = ServiceReferenceHolder.getInstance()
+                    .getAPIManagerConfigurationService()
+                    .getAPIManagerConfiguration().isEnterpriseSubscriptionEnabled();
+            Subscription subscription = null;
+
+            if (isSelfSubscriptionEnabled && isEnterpriseSubscriptionEnabled) {
+                subscription = getIndividualSubscription(consumer, appContext, appVersion, connection);
+                // If there is an individual subscription proceed with it. Try enterprise subscription otherwise
+
+                if(subscription != null){
```
Comment (sajithab) [inline components/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/dao/AppMDAO.java:5702]: Formatting...

File: components/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/dao/AppMDAO.java [hunk 16]
```diff
@@ -5687,22 +5687,50 @@ public APIKeyValidationInfoDTO getApplicationData(String appContext, String appV
...
-            Subscription subscription = getIndividualSubscription(consumer, appContext, appVersion, connection);
+            boolean isSelfSubscriptionEnabled = ServiceReferenceHolder.getInstance()
+                    .getAPIManagerConfigurationService()
+                    .getAPIManagerConfiguration().isSelfSubscriptionEnabled();
+            boolean isEnterpriseSubscriptionEnabled = ServiceReferenceHolder.getInstance()
+                    .getAPIManagerConfigurationService()
+                    .getAPIManagerConfiguration().isEnterpriseSubscriptionEnabled();
+            Subscription subscription = null;
+
+            if (isSelfSubscriptionEnabled && isEnterpriseSubscriptionEnabled) {
```
Comment (sajithab) [inline components/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/dao/AppMDAO.java:5698]: Move to private method. L5698 to L5730

Comment (sajithab) [inline components/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/dao/AppMDAO.java:5698]:
```
if (isSelfSubscriptionEnabled) {
   ...
   if(isEnterpriseSubscriptionEnabled) {
   ...
   }
} else if (!isSelfSubscriptionEnabled) {
...
} else {
...
}
```

File: components/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/dao/AppMDAO.java [hunk 17]
```diff
@@ -5710,44 +5738,47 @@ public APIKeyValidationInfoDTO getApplicationData(String appContext, String appV
...
-		} catch (SQLException e) {
-			handleException("Error while getting application data for : " + appContext + "-" +
-					appVersion, e);
-			return null;
-		} finally {
-			APIMgtDBUtil.closeAllConnections(null, connection, null);
-		}
-	}
+            // Get App info from the database.
+            WebApp webApp = getWebApp(appContext, appVersion);
```
Comment (jenananthan) [inline components/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/dao/AppMDAO.java:5742]: use the same connection

File: components/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/dao/AppMDAO.java [hunk 18]
```diff
@@ -5710,44 +5738,47 @@ public APIKeyValidationInfoDTO getApplicationData(String appContext, String appV
...
+
+            info.setApiName(webApp.getId().getApiName());
+            info.setApiVersion(webApp.getId().getVersion());
+            info.setApiPublisher(webApp.getId().getProviderName());
+            info.setTier(AppMConstants.UNLIMITED_TIER);
+            info.setContext(appContext);
+            info.setLogoutURL(webApp.getLogoutURL());
+
+            if(isEnterpriseSubscriptionEnabled) {
```
Comment (sajithab) [inline components/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/dao/AppMDAO.java]: No need this if. Can move L5753 to below if block.

File: components/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/dao/AppMDAO.java [hunk 19]
```diff
@@ -5710,44 +5738,47 @@ public APIKeyValidationInfoDTO getApplicationData(String appContext, String appV
...
+                    return info;
+                }
+            }
+            // Good to go !
+            info.setAuthorized(true);
+            return info;
+        } catch (SQLException e) {
+            handleException("Error while getting application data for : " + appContext + "-" +
+                                    appVersion, e);
+            return null;
```
Comment (jenananthan) [inline components/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/dao/AppMDAO.java:5777]: no need to return null

File: components/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/dao/AppMDAO.java [hunk 20]
```diff
@@ -5710,44 +5738,47 @@ public APIKeyValidationInfoDTO getApplicationData(String appContext, String appV
...
+            info.setTier(AppMConstants.UNLIMITED_TIER);
+            info.setContext(appContext);
+            info.setLogoutURL(webApp.getLogoutURL());
+
+            if(isEnterpriseSubscriptionEnabled) {
+                // Set trusted IDPs.
+                info.setTrustedIdp(JSONValue.toJSONString(subscription.getTrustedIdps()));
+            }
+
+            if (isSelfSubscriptionEnabled || isEnterpriseSubscriptionEnabled) {
```
Comment (jenananthan) [inline components/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/dao/AppMDAO.java:5756]: return info object in one place

**Pull request merged**